File: DiabloCharacterTracker.Server/Services/ItemAffixServices/ItemAffixService.cs
using DiabloCharacterTracker.Server.Data;
using DiabloCharacterTracker.Server.DTOs;
using DiabloCharacterTracker.Server.Requests.AddRequests;
using Microsoft.EntityFrameworkCore;

namespace DiabloCharacterTracker.Server.Services.ItemAffixServices;

public class ItemAffixService : IItemAffixService
{
    private readonly IDbContextFactory<DiabloDbContext> dbContextFactory;
    public ItemAffixService(IDbContextFactory<DiabloDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }
    public async Task<int> AddItemAffix(AddItemAffixRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var possibleItemAffix = await context.ItemAffixes
            .Where(x => x.AffixId == request.AffixId)
            .Where(x => x.ItemId == request.ItemId)
            .FirstOrDefaultAsync();

        if (possibleItemAffix != null)
        {
            return 0;
        }

        ItemAffix newItemAffix = new ItemAffix()
        {
            AffixId = request.AffixId,
            ItemId = request.ItemId,
        };

        if (newItemAffix == null)
        {
            return 0;
        }

        await context.ItemAffixes.AddAsync(newItemAffix);
        await context.SaveChangesAsync();
        return newItemAffix.Id;
    }

    public async Task<bool> DeleteItemAffix(int itemAffixId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var itemAffixToDelete = await context.ItemAffixes
            .Where(x => x.Id == itemAffixId)
            .FirstOrDefaultAsync();

        if (itemAffixToDelete == null)
        {
            return false;
        }

        context.Remove(itemAffixToDelete);
        await context.SaveChangesAsync();
        return true;
    }

    public async Task<List<AffixDTO>> GetAllAffixesByItem(int itemId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var affixes = await context.ItemAffixes
            .Include(x => x.Affix)
            .Include(x => x.Item)
            .Where(x => x.ItemId == itemId)
            .ToListAsync();

        if (affixes == null)
        {
            return new List<AffixDTO>();
        }

        return affixes.Select(x => x.Affix!.ToDTO()).ToList();
    }

    public async Task<List<ItemDTO>> GetAllItemsByAffix(int affixId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var items = await context.ItemAffixes
            .Include(x => x.Affix)
            .Include(x => x.Item)
            .Where(x => x.AffixId == affixId)
            .ToListAsync();

        if (items == null)
        {
            return new List<ItemDTO>();
        }

        return items.Select(x => x.Item!.ToDTO()).ToList(); 
    }

    public async Task<ItemAffixDTO> GetItemAffixById(int itemAffixId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var itemAffix = await context.ItemAffixes
            .Where(x => x.Id == itemAffixId)
            .FirstOrDefaultAsync();

        if (itemAffix == null)
        {
            return new ItemAffixDTO();
        }

        return itemAffix.ToDTO();
    }
}

File: diablocharactertracker.client/src/ApiServices/ItemAffixService.ts
import { Item } from "@/Data/DTOs/Item";
import { Affix } from "../Data/DTOs/Affix";
import axios from "axios";
import { ItemAffix } from "@/Data/DTOs/ItemAffix";
import { AddItemAffixRequest } from "@/Data/Requests/AddRequests/AddItemAffixRequest";

export const ItemAffixService = {
  GetAllAffixesByItem: async () => {
    try {
      const response = await axios.get<Affix[]>(
        `${import.meta.env.VITE_URL}/ItemAffix/getallaffixesforitem`
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get all affexes by item");
      throw error;
    }
  },
  GetAllItemsByAffix: async () => {
    try {
      const response = await axios.get<Item[]>(
        `${import.meta.env.VITE_URL}/ItemAffix/getallitemsbyaffix`
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get all items by affix");
      throw error;
    }
  },
  GetItemAffixById: async (itemAffixId: number) => {
    try {
      const response = await axios.get<ItemAffix>(
        `${import.meta.env.VITE_URL}/ItemAffix/getitemaffixbyid`,
        {
          params: {
            itemAffixId,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get item affix by id");
      throw error;
    }
  },
  AddItemAffix: async (addItemAffixRequest: AddItemAffixRequest) => {
    try {
      const response = await axios.post<number>(
        `${import.meta.env.VITE_URL}/ItemAffix/additemaffix`,
        addItemAffixRequest,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to add affix to item");
      throw error;
    }
  },
};
